### ISSUE TITLE
Doesn't work in MacOS Mohave

### ISSUE THREAD

**Issue by PVGrad** [yousseb/meld#70]
I run it
```
meld /etc/php.ini.default-previous /etc/php.ini.default-previous\~orig
/Applications/Meld.app/Contents/MacOS
```
And nothing happens - only icon of meld is showing, but no window
Then I press ctrl+c to terminate process and this is the output:
```
Traceback (most recent call last):
  File "/usr/local/bin/meld", line 47, in <module>
    status = subprocess.call([MELDPATH] + arglist, env=environment)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 522, in call
    return Popen(*popenargs, **kwargs).wait()
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 1384, in wait
    pid, sts = _eintr_retry_call(os.waitpid, self.pid, 0)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 476, in _eintr_retry_call
    return func(*args)
KeyboardInterrupt
```

**Comment by PVGrad** [comment]
As i understood - there were my local problems with python2.7 after system update to Mohave.
So issue can be closed.
Thanks

**Comment by zzju** [comment]
@PVGrad So how to resolve this problem? 
macOS 10.14
python --version
Python 2.7.10

many thanks.

**Comment by zzju** [comment]
installed python 3, and re-installed meld 3.19, meld works again.

**Comment by fafafariba** [comment]
I have python 3 installed as well as meld 3.19, but still having the same problem as OP.

**Comment by munepi** [comment]
Please try to run the following Bash script `_clean-meld-settings.sh` to cleanup your some settings of Meld. 

``` bash:_clean-meld-settings.sh
#!/bin/bash

cd ${HOME}
rm -rf ./.local/share/meld
rm -f ./Library/Preferences/org.gnome.meld.plist
rm -rf "./Library/Saved Application State/org.gnome.meld.savedState/"
```

**Comment by munepi** [comment]
Today morning (UCT+9 🇯🇵), I also have posted how to use/change custom-font with Meld plist settings in Issue [Can't change font #38](https://github.com/yousseb/meld/issues/38). 
The plist file `~//Library/Preferences/org.gnome.meld.plist` is just the one. 
Please check and try the post if necessary.

**Comment by fafafariba** [comment]
> Please try to run the following Bash script `_clean-meld-settings.sh` to cleanup your some settings of Meld.
> 
> ```shell
> #!/bin/bash
> 
> cd ${HOME}
> rm -rf ./.local/share/meld
> rm -f ./Library/Preferences/org.gnome.meld.plist
> rm -rf "./Library/Saved Application State/org.gnome.meld.savedState/"
> ```

This worked for me, thank you.

**Comment by ppannuto** [comment]
I've got approximately the same output running Meld (python version 2.7.15, OS X 10.14). Cleaning settings did not change anything for me:

```
/Applications/Meld.app/Contents/MacOS
^CTraceback (most recent call last):
  File "/Applications/Meld.app/Contents/MacOS/Meld", line 47, in <module>
    status = subprocess.call([MELDPATH] + arglist, env=environment)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 172, in call
    return Popen(*popenargs, **kwargs).wait()
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 1099, in wait
    pid, sts = _eintr_retry_call(os.waitpid, self.pid, 0)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 125, in _eintr_retry_call
    return func(*args)
KeyboardInterrupt
```

Anything else I can try?

**Comment by MaxBenChrist** [comment]
> Please try to run the following Bash script `_clean-meld-settings.sh` to cleanup your some settings of Meld.
> 
> ```shell
> #!/bin/bash
> 
> cd ${HOME}
> rm -rf ./.local/share/meld
> rm -f ./Library/Preferences/org.gnome.meld.plist
> rm -rf "./Library/Saved Application State/org.gnome.meld.savedState/"
> ```

this worked for me as well, thank you! 👍

**Comment by JerryShih** [comment]
> > Please try to run the following Bash script `_clean-meld-settings.sh` to cleanup your some settings of Meld.
> > ```shell
> > #!/bin/bash
> > 
> > cd ${HOME}
> > rm -rf ./.local/share/meld
> > rm -f ./Library/Preferences/org.gnome.meld.plist
> > rm -rf "./Library/Saved Application State/org.gnome.meld.savedState/"
> > ```
> 
> this worked for me as well, thank you! 👍

It worked only 1 time after cleaning all setting. I still only see an icon for the second and the after runs.

**Comment by munepi** [comment]
@JerryShih 
That is an unofficial, bad known-how :-( , but please try the following three steps. 

 1. copy the original Mac-port starter script to `Meld.distro` below.
    ```bash
    $ cp -a /Applications/Meld.app/Contents/MacOS/Meld /Applications/Meld.app/Contents/MacOS/Meld.distro
    ```
2. make a wrapper `/Applications/Meld.app/Contents/MacOS/Meld` of the `Meld.distro` below
    ```bash
    #!/bin/bash
     
    ## 
    # $ sudo xattr -r -d com.apple.quarantine /Applications/Meld.app/
     
    ## standard plist settings: OK
    # ~/Library/Preferences/org.gnome.meld.plist
    # $ defaults read org.gnome.meld
    # $ defaults write org.gnome.meld "/org/gnome/meld/use-system-font" 0
    # $ defaults write org.gnome.meld "/org/gnome/meld/custom-font" "Ricty, 18"
     
    ##FIXME: ad-hoc cleanup Meld settings
    (cd ${HOME}
    rm -rf ./.local/share/meld
    #rm -f ./Library/Preferences/org.gnome.meld.plist
    rm -rf "./Library/Saved Application State/org.gnome.meld.savedState/"
    )
     
    # export G_ENABLE_DIAGNOSTIC=1 # 0 or 1
    # export GTK_THEME=Adwaita # Adwaita or Adwaita:dark
    # export GTK_DEBUG=interactive
     
    logger "/Applications/Meld.app/Contents/MacOS/Meld.distro"
    exec "/Applications/Meld.app/Contents/MacOS/Meld.distro" $@
    ````
3. Then, `$ chmod +x /Applications/Meld.app/Contents/MacOS/Meld`

I remark that 
 * you are available some GTK-related environment variables inside the above wrapper `Meld`: `G_ENABLE_DIAGNOSTIC`, `GTK_THEME`, `GTK_DEBUG`. 
 * I also have posted how to use/change custom-font with Meld plist settings in Issue [Can't change font #38](https://github.com/yousseb/meld/issues/38#issuecomment-426340177). 
The plist file `~//Library/Preferences/org.gnome.meld.plist` is just the one. 
Please check and try the post if necessary.

**Comment by JerryShih** [comment]
> @JerryShih
> That is an unofficial, bad know-how :-( , but please try the following three steps.
> 
> 1. move the original Mac-port starter script to `Meld.distro` below.
>    ```shell
>    $ mv /Applications/Meld.app/Contents/MacOS/Meld /Applications/Meld.app/Contents/MacOS/Meld.distro
>    ```
> 2. make a wrapper `/Applications/Meld.app/Contents/MacOS/Meld` of the `Meld.distro` below
>    ```shell
>    #!/bin/bash
>     
>    ## 
>    # $ sudo xattr -r -d com.apple.quarantine /Applications/Meld.app/
>     
>    ## standard plist settings: OK
>    # ~/Library/Preferences/org.gnome.meld.plist
>    # $ defaults read org.gnome.meld
>    # $ defaults write org.gnome.meld "/org/gnome/meld/use-system-font" 0
>    # $ defaults write org.gnome.meld "/org/gnome/meld/custom-font" "Ricty, 18"
>     
>    ##FIXME: ad-hoc cleanup Meld settings
>    (cd ${HOME}
>    rm -rf ./.local/share/meld
>    #rm -f ./Library/Preferences/org.gnome.meld.plist
>    rm -rf "./Library/Saved Application State/org.gnome.meld.savedState/"
>    )
>     
>    # export G_ENABLE_DIAGNOSTIC=1 # 0 or 1
>    # export GTK_THEME=Adwaita # Adwaita or Adwaita:dark
>    # export GTK_DEBUG=interactive
>     
>    logger "/Applications/Meld.app/Contents/MacOS/Meld.distro"
>    exec "/Applications/Meld.app/Contents/MacOS/Meld.distro" $@
>    ```
> 3. Then, `$ chmod +x /Applications/Meld.app/Contents/MacOS/Meld`
> 
> I remark that
> 
> * you are available some GTK-related environment variables inside the above wrapper `Meld`: `G_ENABLE_DIAGNOSTIC`, `GTK_THEME`, `GTK_DEBUG`.
> * you are also available some tips commented in the above wrapper `Meld` if necessary.
> * I also have posted how to use/change custom-font with Meld plist settings in Issue [Can't change font #38](https://github.com/yousseb/meld/issues/38#issuecomment-426340177).
>   The plist file `~//Library/Preferences/org.gnome.meld.plist` is just the one.
>   Please check and try the post if necessary.

It just a wrapper to clean the setting before calling meld. What's the root cause for this problem?
Do all gtk-base gui tools have the same problem?

**Comment by yousseb** [comment]
Not all gtk-based, just the pygtk-based require such wrapper. 

And I apologize for not being able to help as efficiently on this thread (and others), but I promise that once I have some spare time on my hands + having collected some tricks and fixes from the people that comment on those issues, I'll create a new release that would include the fixes in one nifty release.

**Comment by munepi** [comment]
Hi @yousseb , Thank you for your information and follow-up. 
I exactly could not trace that problem around pygtk-something :-( So, I have made such ad-hoc wrapper. 
I'm looking forward to a new release!

**Comment by yousseb** [comment]
@munepi thank you for trying (and helping others get their cases resolved). I wish there was some means to tell the generated binary the path to look for libraries in a cleaner way to get rid of the python wrapper that I had to make. That on its own would have solved other issues on my list here. 

But anyway, I guess I'll be able to put sometime into this within a month or so. Or at least I hope..

**Comment by gopeshkx** [comment]
> #!/bin/bash cd ${HOME} rm -rf ./.local/share/meld rm -f ./Library/Preferences/org.gnome.meld.plist rm -rf "./Library/Saved Application State/org.gnome.meld.savedState/"

Work for me as well. 👍

**Comment by chaosless** [comment]
cd ${HOME}
rm -rf ./.local/share/meld
rm -f ./Library/Preferences/org.gnome.meld.plist
rm -rf "./Library/Saved Application State/org.gnome.meld.savedState/"

**Comment by decentralizeduser** [comment]
> ```shell
> rm -rf "./Library/Saved Application State/org.gnome.meld.savedState/"
> ```

Nice .. thanks for that :-)

**Comment by oxoocoffee** [comment]
I have same issue on Mojave with 3.19.0

**Comment by bbqchickenrobot** [comment]
The fix kind of works for me but the mouse isn't accurate when trying to manipulate the UI elements.  I have to go about a quarter of an inch above the UI element for the click event to register.  Hopefully this will be fixed sooner than later!

**Comment by decentralizeduser** [comment]
> The fix kind of works for me but the mouse isn't accurate when trying to manipulate the UI elements. I have to go about a quarter of an inch above the UI element for the click event to register. Hopefully this will be fixed sooner than later!

I also get that, but only when resizing the application window after opening, if I don't change the size of the application window I don't get that problem.

**Comment by dluc** [comment]
> Please try to run the following Bash script `_clean-meld-settings.sh` to cleanup your some settings of Meld.
> 
> ```shell
> #!/bin/bash
> 
> cd ${HOME}
> rm -rf ./.local/share/meld
> rm -f ./Library/Preferences/org.gnome.meld.plist
> rm -rf "./Library/Saved Application State/org.gnome.meld.savedState/"
> ```

Thanks, this fixed it, without having to upgrade Python 🏁 

Please add this to the release notes :-)

**Comment by SemihGk** [comment]
> Please try to run the following Bash script `_clean-meld-settings.sh` to cleanup your some settings of Meld.
> 
> ```shell
> #!/bin/bash
> 
> cd ${HOME}
> rm -rf ./.local/share/meld
> rm -f ./Library/Preferences/org.gnome.meld.plist
> rm -rf "./Library/Saved Application State/org.gnome.meld.savedState/"
> ```

Removing some meld files temporarily fix the issue. So, I wanted to try the wrapper solution as munepi posted above, but no luck I am receiving "Application cannot be opened because its executable is missing" error. Probably, I am doing something wrong at step2 of the wrapper solution. Anyway, I also look forward to waiting for an update for this issue. Thanks a lot for your help.

**Comment by dluc** [comment]
Using 3.19.0 and Meld stopped working again. Repeating the steps above fixed the issue one more time, but apparently there's a bug leading to the broken state at some point

**Comment by yousseb** [comment]
Hey guys, could you please run the `_clean-meld-settings.sh` script by @munepi  in this thread then give this new build a shot?

https://github.com/yousseb/meld/releases/tag/os-11

**Comment by preinking** [comment]
Running 3.19.2 (r1) but the font size is 18pt and cannot be changed as this leads to a crash

**Comment by ira-bv** [comment]
tried newest version, no go.

```
 ↳ meld --version
 meld 3.19.2
```

But maybe this is a clue, Yousef?

```
 ↳ /Applications/Meld.app/Contents/MacOS/Meld-bin
 Couldn't set the locale: unsupported locale setting; falling back to 'C' locale

 ** (process:26632): WARNING **: 14:32:29.078: Failed to load shared library 
 '/Users/youssef/gtk/inst/lib/libglib-2.0.0.dylib' referenced by the typelib: 
 dlopen(/Users/youssef/gtk/inst/lib/libglib-2.0.0.dylib, 9): image not found

 ** (process:26632): WARNING **: 14:32:29.078: Failed to load shared library 
 '/Users/youssef/gtk/inst/lib/libgobject-2.0.0.dylib' referenced by the typelib: 
 dlopen(/Users/youssef/gtk/inst/lib/libgobject-2.0.0.dylib, 9): image not found
```

**Comment by yousseb** [comment]
Trying to figure this out. But no, it’s not those paths.

The Meld starter python script sets the DYLD_LIBRARY_PATH which tells
dlopen where to look. In fact, this is the only reason why the script
exists and I’ve been looking for a way around this forever. This is deep
inside gopject-introspection python compiler. The only recommendation that
I have is don’t start Meld-bin directly until I find a solution to telling
gi to use relative path.

Alternatively, I could use some help in telling macOS not save state on
crashes. Looks like this is what is happening.

**Comment by ira-bv** [comment]
Seems to work and repetitively :) Thanks!! 👍

**Comment by cmrust** [comment]
Works repeatedly for me too! Thanks!

**Comment by cyberdemon8** [comment]
Works!!! Thanks! 👍

**Comment by preinking** [comment]
Works for me as well Thanks! 👍

**Comment by MaxBenChrist** [comment]
Thanks!

is there already a release scheduled for homebrew/cask?

**Comment by SemihGk** [comment]
I tried the latest build. When I merge some conflicts, I constantly receive:
```There was a problem opening the file “somepath/somefile_LOCAL.java”.```
```There was a problem opening the file “somepath/somefile_BASE.java”.```
```There was a problem opening the file “somepath/somefile_REMOTE.java”.```

I checked the path. Those 3 files exist. Somehow, meld is unable to open those files. Is there any guess why it is happening? Stable build does not have this path problem if I use clean-meld.settings.sh script before each file conflict.

**Comment by yousseb** [comment]
@Semih thanks for the report. Let’s open a separate issue for it and please
share your .gitconfig in the issue as well.

**Comment by SemihGk** [comment]
Sure. I will open a separate ticket. Thank you for the response.

**Comment by knumat** [comment]
@yousseb Thanks for the updated build!  It starts up fine for me on macOS 10.14.1 (Mojave).  Hurray!
> https://github.com/yousseb/meld/releases/tag/osx-11

I am seeing a similar issue to @bbqchickenrobot where I have to click a bit above the buttons when I first start it up.  However, once I resize the window, everything seems to be fine (which is the opposite behavior that @decentralizeduser saw).  After some experimentation, it appears that showing/hiding tab bar does not re-layout the contents of the window, but resizing the window does.  I probably notice this more than more users because I enabled Settings -> Dock -> Prefer Tabs When Opening Documents -> Always.

Reproduction steps:

1. Start Meld.
2. Select View -> Show Tab Bar
3. Move the mouse over a button and notice that the hover effect is doesn't line up with the top/bottom edges (mouse needs to be 1/4inch higher).
4. Resize the window.
5. Move the mouse over a button and notice that the hover effect properly lines up with the top/bottom edges.
6. Select View -> Hide Tab Bar
7. Move the mouse over a button and notice that the hover effect is doesn't line up with the top/bottom edges (mouse needs to be 1/4inch lower).
8. Resize the window.
9. Move the mouse over a button and notice that the hover effect properly lines up with the top/bottom edges.

Let me know if you would like me to file a separate issue.

**Comment by nobodo** [comment]
That mouse disalignment issue is not limited to Meld only. There is at least one third party program (GLM 3 by Genelec) that has the same issue which has the same workaround by resizing the window. I have tried to isolate the problem and it could be related to the tabbed layout, but I'm not sure which library to blaim.

**Comment by oliveratutexas** [comment]
I installed via cask and am still have some of these problems, the script
```
#!/bin/bash

cd ${HOME}
rm -rf ./.local/share/meld
rm -f ./Library/Preferences/org.gnome.meld.plist
rm -rf "./Library/Saved Application State/org.gnome.meld.savedState/"
``` 
worked on the first run for me, not sure what else to do. For git mergetool config I had 
```
[diff]
  tool = meld
[difftool]
  prompt = false
[difftool "meld"]
  trustExitCode = true
  cmd =  meld --args $LOCAL $REMOTE
[merge]
  tool = meld
[mergetool]
  prompt = false
[mergetool "meld"]
  trustExitCode = true
  cmd =  echo "$LOCAL" && echo "$BASE" && echo $REMOTE && meld --auto-merge $LOCAL $BASE $REMOTE --output=$MERGED
```
tried on MacOS

**Comment by oliveratutexas** [comment]
> The following will ensure that the parameters are always fully qualified paths. It used to work as you had it on the last version, but the latest builds sometimes don't. This is how I fixed it.
> 
> ```
> [diff]
>   tool = meld
> [difftool "meld"]
>   cmd = open -W -n -a Meld --args $(cd $(dirname \"$LOCAL\") && pwd)/$(basename \"$LOCAL\") $(cd $(dirname \"$REMOTE\") && pwd)/$(basename \"$REMOTE\")
>   trustExitCode = true
> [merge]
>   tool = meld
> [mergetool "meld"]
>   cmd = open -W -n -a Meld --args --auto-merge  $(cd $(dirname \"$LOCAL\") && pwd)/$(basename \"$LOCAL\") $(cd $(dirname \"$BASE\") && pwd)/$(basename \"$BASE\") $(cd $(dirname \"$REMOTE\") && pwd)/$(basename \"$REMOTE\") --output=$(cd $(dirname \"$MERGED\") && pwd)/$(basename \"$MERGED\")
>   trustExitCode = true
> ```

That helped, thank you!

**Comment by fiendish** [comment]
3.19.2-r6 still exhibits this issue if I ctrl+C meld from the terminal, but when I run meld I get the Meld menu bar. If I go to `Meld->Quit Meld` then it works again after that.

**Comment by bmusat** [comment]
When you run 'meld' from the command line and **ctrl-C** it, meld does not clean up after itself.  You need to delete the `~/Library/Saved Application State/org.gnome.meld.savedState/` before restarting.  This directory is deleted if you `Meld->Quit Meld` which is why that works.

I modified the `/usr/local/Caskroom/meld/3.19.2-r6,osx-15/meld.wrapper.sh` script installed by brew to the following:

> #!/bin/sh
> rm -rf ~/Library/Saved\ Application\ State/org.gnome.meld.savedState
> exec '/Applications/Meld.app/Contents/MacOS/Meld' "$@"

and now Meld works if you invoke it from the command line every time.  

I don't think there is _any need_ to remove these other two directories or files since they only contain your preferences and recently compared files:
> rm -rf ./.local/share/meld
> rm -f ./Library/Preferences/org.gnome.meld.plist

**Comment by p1nox** [comment]
Thanks @bmusat , worked for me 💯 🎉

**Comment by bmusat** [comment]
Also wanted to note that if you are still having problems then you might have an old Meld process lying around.  To see if you do:
> pgrep -fi meld

will list the PIDs and then:
> pkill -fi meld

will kill them off..  and the `meld.wrapper.sh` script in my previous post above should clean up the rest.  I suppose you could add the `pkill` in that script as well, just to be sure, but I don't think it's necessary.

**Comment by rmeffan** [comment]
Running meld `3.19.2-r6,osx-15`, and I ran into the problem described with this issue. It's possible that it either started after upgrading to Mojave from ElCap, but it's also possible that I have used `CTRL-C` at some point to kill meld when it wasn't responding.

I ran the cleanup script that was recommended in this comment, and it resolved the issue. https://github.com/yousseb/meld/issues/70#issuecomment-425686670

If the problem recurs, I will try modifying the wrapper as described here:
https://github.com/yousseb/meld/issues/70#issuecomment-479199105

Thanks, all!

**Comment by ffloimair** [comment]
I'm having the same issue. Have to run the 3 additional steps above manually each time before starting Meld. Otherwise the window won't show.

**Comment by cscetbon** [comment]
> When you run 'meld' from the command line and **ctrl-C** it, meld does not clean up after itself. You need to delete the `~/Library/Saved Application State/org.gnome.meld.savedState/` before restarting. This directory is deleted if you `Meld->Quit Meld` which is why that works.

@bmusat even if I quit I get this issue, so I think it's an issue in both cases. Thanks for the tip but We should add it to the brew formula until it's definitely fixed ...

**Comment by terriblefire** [comment]
Also seeing the issue on MacOSX Mojave

**Comment by slmingol** [comment]
On Mojave 10.14.4 and meld: 3.19.2-r6,osx-15. Above issue(s) are still occurring. I had to resort to using the same method shown by others. I tweaked it as follows just b/c I like to use the dark theme. I also removed the `cd` and opted for fully qualified paths vs. relative to move things along.

```
$ more ~/bin/meld
#!/bin/bash

rm -rf ~/.local/share/meld \
    ~/Library/Preferences/org.gnome.meld.plist \
    "~/Library/Saved Application State/org.gnome.meld.savedState/"

GTK_THEME=Adwaita:dark meld $*
```

**Comment by fredleger** [comment]
got the feeling this error raise when you are mixing beetween fullscreen native app in macosx and desktop opened apps. meld would try to open on desktop even if you are in a fullscreen app (like iterm2 in my case).

Does that make any sense ?

**Comment by slmingol** [comment]
I think you were suggesting trying to suggest that under the View pulldown you can check/uncheck Fullscreen? I tried toggling that and it made no difference b/w fullscreen and desktop.

**Comment by fredleger** [comment]
No i mean using OS native fullscreen app support (available on all apps). Clicking on the green button on the window and using exposé to navigate through fullscreen apps:

<img width="1392" alt="Capture d’écran 2019-09-25 à 15 14 43" src="https://user-images.githubusercontent.com/2778741/65604296-7fe8b380-dfa7-11e9-8c14-8b0656ac41fe.png">
<img width="1091" alt="Capture d’écran 2019-09-25 à 15 16 14" src="https://user-images.githubusercontent.com/2778741/65604297-80814a00-dfa7-11e9-8516-befe7d0caf24.png">

**Comment by y0y0z** [comment]
still happens with meld version 3.19.2-r6,osx-15 on macos version 10.14.6:

```bash
$ brew cask info meld | head -n1
meld: 3.19.2-r6,osx-15

$ sw_vers
ProductName:	Mac OS X
ProductVersion:	10.14.6
BuildVersion:	18G103
```

solved it using @bmusat [solution][1].
i made a [pull-request to fix the wrapper provided with the brew formula][2]

[1]: https://github.com/yousseb/meld/issues/70#issuecomment-479199105
[2]: https://github.com/Homebrew/homebrew-cask/pull/69864

**Comment by Napoleon-BlownApart** [comment]
I've been using the suggested script to clean the stored data before each use of Meld.  Though this works, it's a bit of a pain, having to go to the terminal each time.

For fix (hack) this, I did the following (using my existing install):
1) Renamed `/Applications/Meld.app/Content/MacOS/Meld` to  `/Applications/Meld.app/Content/MacOS/Meld.py`
2) Created a **new** file, with execute permissions `/Applications/Meld.app/Content/MacOS/Meld`
3) Put the three file/dir delete commands into this new file, along with a bash shibang at the beginning and an statement to execute Meld.py (see step 1).

The new `Meld` script:  


    #!/bin/bash
    cd ${HOME}
    rm -rf ./.local/share/meld
    rm -f ./Library/Preferences/org.gnome.meld.plist
    rm -rf "./Library/Saved Application State/org.gnome.meld.savedState/"
   
    cd /Applications/Meld.app/Contents/MacOS/
    ./Meld.py


Now I can use Meld without needing to manually clean the app data each time.  :)
(I also noticed that the first two removes are not always necessary.)

**Comment by knumat** [comment]
@Napoleon-BlownApart I haven't been having issues since installing this version:
https://github.com/yousseb/meld/releases/tag/osx-11
If you are still having issues, your script seems like a possible workaround.  You might want to add a `"$@"` to allow command line arguments to be passed through so Sourcetree, git difftool, etc are functional.  Make sure to save a copy of your final script somewhere else, since it will be removed when you upgrade Meld.

**Comment by Napoleon-BlownApart** [comment]
Ok, thanks for the tip.  I've updated and essentially it works.
But...  How do I tell Catalina that Meld is allowed to access my documents?  (I have the same problem with the latest version of DIA.)

**Comment by twh270** [comment]
This has been an interesting journey (there and over yonder and finally back again). I'm on [Catalina 10.15.4, Meld 3.19.2, git 2.26.0] and after reading and trying a lot of things, here's what I have working. 

First up, the relevant bit of my ~/.gitconfig
`[merge]
        tool = meld
[diff]
        tool = meld
[difftool]
        prompt = false
[mergetool]
        keeptemporaries = false
        keepbackups = false
        prompt = false
        keepBackup = false
[difftool "meld"]
        trustExitCode = false
        cmd = meld \"$LOCAL\" \"$REMOTE\"
[mergetool "meld"]
        trustExitCode = false
        cmd = meld --auto-merge \"$LOCAL\" \"$BASE\" \"$REMOTE\" --output=\"$MERGED\"

`

And the kludge de resistance, the script that makes it work, ~/bin/meld

`
#!/usr/bin/python

import sys
import os
import subprocess
import shutil

MELDPATH = "/Applications/Meld.app/Contents/MacOS/Meld"

home = os.path.expanduser("~")
ignore_errors = True
shutil.rmtree(home + '/.local/share/meld', ignore_errors)
if os.path.exists(home + '/Library/Preferences/org.gnome.meld.plist'):  
    os.remove(home + '/Library/Preferences/org.gnome.meld.plist')
shutil.rmtree(home + '/Library/Saved Application State/org.gnome.meld.savedState', ignore_errors)

userArgs = []
for arg in sys.argv[1:]:
    if arg.startswith('--output'):
        filename = arg.split('=')[1]
        newArg = '--output=' + os.path.abspath(filename)
    elif arg.startswith('--auto-merge'):
        newArg = arg
    elif arg.startswith('-'):
        newArg = os.path.abspath(arg)
    else:
        newArg = os.path.abspath(arg)
    userArgs.append(newArg)

argsArray = ['open', '-W', '-a', MELDPATH, '--args'] + userArgs

# (userArgs) # (sys.argv[1:])
print argsArray

try:
    p = subprocess.call(argsArray)
except:
    pass
`

If you don't have Meld working for you, perhaps this will help.

**Comment by Napoleon-BlownApart** [comment]
@twh270 : I tried your suggestion, but I'm still getting the MacOS Catalina permission denial.  Whilst I don't think it makes a difference here, I'm running Catalina 10.15.3 here.

@knumat :  Would you have any suggestions on how to make the MacOS permission denials go away?

**Comment by fredleger** [comment]
I guess it's because the new perms system in MAcOSx. Didn't try it but
would suggest to go to security and privacy > Confidentiality and give full
disk access to Meld. I noticed all QT apps are doing the same since
Cataline (Gimp for example).

-- 
[image: webofmars | build-and-run]
Frederic Leger

fondateur | webofmars | build-and-run
M:  06.52.77.53.54
E:  contact@webofmars.com
webofmars.com
<https://twitter.com/webofmars>
<https://www.linkedin.com/company/webofmars/> <https://github.com/webofmars>
<https://www.youtube.com/channel/UCi77lDIsszaryN0flTS9EOw>
<https://www.build-and-run.fr/>
P.S. During business hours, focus is my work philosophy. So to fully serve
my customers, I do not consult emails during the day. Please send me an SMS
if anything urgent otherwise my reply might take a few workdays

**Comment by Napoleon-BlownApart** [comment]
@fredleger : Thanks for taking a look.
I've already done all that, and given permission to XQuartz and others (whilst trying to get `DIA` to work).
I haven't figured it out yet (due to time constraints).  For the moment I've reverted back to 3.19.0.

**Comment by knumat** [comment]
I had been having the same issue as @fiendish where a Ctrl+C still left Meld in a bad state, but a Meld -> Quit Meld fixed it.  However, I just installed the latest release from homebrew and I can no longer create this error.  I'm not using any extra scripts like @twh270 and @Napoleon-BlownApart said they needed to use.

My install directions:
- Run `brew cask uninstall meld`
- Run `brew cask install meld`
- Run `meld`
- Click "Cancel" on the security prompt (do not select "Move to trash")
- Open Settings -> Security & Privacy -> General and select Open Anyway and then Open
- Run `meld`

Here are the relevant versions I am running:
```
~$ sw_vers
ProductName:	Mac OS X
ProductVersion:	10.15.4
BuildVersion:	19E287
~$ meld --version
meld 3.21.0.osx3
~$ ls -l `which meld`
lrwxr-xr-x  1 knumat  admin  57 Apr 15 13:16 /usr/local/bin/meld -> /usr/local/Caskroom/meld/3.21.0-r3,osx-19/meld.wrapper.sh
~$ python3 --version
Python 3.7.2
~$ python --version
Python 2.7.15
```

Nothing special in my .gitconfig:
```
[diff]
        tool = meld
[difftool]
        prompt = false
[difftool "meld"]
        trustExitCode = true
        cmd = meld \"$LOCAL\" \"$REMOTE\"
[merge]
        tool = meld
[mergetool]
        prompt = false
        keepBackup = false
[mergetool "meld"]
        trustExitCode = true
        cmd = meld --auto-merge \"$LOCAL\" \"$BASE\" \"$REMOTE\" --output=\"$MERGED\"
```

@Napoleon-BlownApart I'm not an expert on macOS permissions, but I don't think you need to add Full Disk Access for Meld.  I only saw two prompts to allow SourceTree and Terminal to access my Documents folder.  After that, Meld worked fine, since I only start Meld from within SourceTree or Terminal.  Of course, if you launch Meld in stand-alone mode, you would see an additional prompt.  You can review the list of approved/denied requests under Settings -> Security & Privacy -> Privacy -> Files & Folders.  If you are still having permission denied errors, perhaps it would make sense to open a separate issue to address that question?

**Comment by yousseb** [comment]
@knumat  Thank you - you're the first to confirm the fix as in the release notes for
https://github.com/yousseb/meld/releases/tag/osx-19 :)

**Comment by knumat** [comment]
@yousseb Thanks for fixing this annoying issue!

I also wanted to say thanks for sharing the story about your brother in the release notes.  It sounds like he was a wonderful man.

**Comment by fredleger** [comment]
I am also giving a try to @knumat proposal i will make a feedback on it here

**Comment by llorracc** [comment]
@twh270, thanks for your https://github.com/yousseb/meld/issues/70#issuecomment-612306597, this is what finally got it working for me.  

Two notes:
1. As of 2020-04-18, the first line is an editing error:
        tool = meld
should be deleted (otherwise git difftool gives an error message
1. For users who do not have ~/usr/bin in their path, an alternative is either put your meld python script in /usr/local/bin or put it somewhere else and make a link to it in /usr/local/bin

Thanks again!

**Comment by gibfahn** [comment]
Also no longer seeing this on latest brew install (using Catalina):

```console
$ meld --version
meld 3.21.0.osx3
```

**Comment by slmingol** [comment]
```
$ sw_vers -productVersion
10.14.6
```

```
$ meld --version
meld 3.21.0.osx3
```

Confirmed the issue seems resolved for me as well.

**Comment by seralouk** [comment]
> Please try to run the following Bash script `_clean-meld-settings.sh` to cleanup your some settings of Meld.
> 
> ```shell
> #!/bin/bash
> 
> cd ${HOME}
> rm -rf ./.local/share/meld
> rm -f ./Library/Preferences/org.gnome.meld.plist
> rm -rf "./Library/Saved Application State/org.gnome.meld.savedState/"
> ```

This worked on my macbook 2019, Catalina built-in. Thanks

**Comment by keikoro** [comment]
While the info about `org.gnome.meld.savedState` potentially being the culprit is helpful, you do absolutely _not_ want run `rm -f ./Library/Preferences/org.gnome.meld.plist` if you (might) have added your own text filters to Meld. At least not before making a backup of the file first...
